scripts/build-scenario.ts

```diff
@@ -0,0 +1,73 @@
+// Build a scenario with given ID from CLI args
+
+// Usage: `yarn scenario [ID]`
+// ID = name of scenario script OR 'default' if no ID provided
+
+import { ScenarioBuilder, Scenario } from './content-utils'
+import { promises } from 'fs'
+import { GameWorld } from '../src/game/ContentTypes'
+const { writeFile, mkdir } = promises
+
+const id = process.argv.length >= 3 ? process.argv[2] : 'default'
+
+async function buildScenario(id: string) {
+    try {
+        const builder: ScenarioBuilder = (
+            await import(`./scenarios/${id}/index.ts`)
+        ).builder
+
+        const scenario = builder.run()
+
+        console.log(JSON.stringify(scenario, null, 2))
+        console.log(`✅ Sucessfully built scenario "${id}"`)
+
+        return scenario
+    } catch (e) {
+        throw new Error(`Error: Failed to build scenario "${id}" - ${e}`)
+    }
+}
+
+async function exportScenario(id: string, scenario: Scenario) {
+    // TODO: export scenario to JSON
+    // Build into separate files and store in a output directory.
+    // Create files in a directory, or write to existing if they already have been created.
+
+    type Parts = {
+        [K in keyof GameWorld]: string
+    }
+
+    const scenarioParts: Parts = {
+        cards: 'cards',
+        eventCards: 'event-cards',
+        events: 'events',
+        defaultState: 'default-state',
+        stats: 'stats',
+    }
+
+    // Create the necessary directories if they are missing
+    const outputDir = `${__dirname}/build/${id}`
+    await mkdir(outputDir, { recursive: true })
+
+    await Promise.all(
+        Object.entries(scenarioParts).map(([part, fileName]) =>
+            writeFile(
+                `${outputDir}/${fileName}.json`,
+                JSON.stringify(scenario[part as keyof GameWorld], null, 4),
+            ),
+        ),
+    )
+
+    // TODO: improve the export, verify that it works correcly by diffing JSON output with default scenario
+    // TODO: Export directly to the /src/public/scenarios directory to make it playable in game
+    // TODO: Test export in the game and see if it's playable
+    // TODO: simplify prop references to make output cleaner, since data might be edited manually in other editors - and that the generated prop references will change on every build.
```

Comment:
The current implementation of propRef will remain static as long as the scripted content is static. I would like to discuss the reality of combining manual edits with scripted builds :). This is no simple task and I think for the moment the best way forward is using diff tools on the generated result and inserting the edited data as code in order to maintain consistency.

That said. It is alway possible to namespace the propRef in order to make it easier to avoid name conflicts with manuallly edited cards. Ex. the prefix of the propRef could be used this way. The good thing about most of the formats we have created is that they are mostly very easy to merge as long as you have consistent naming conventions to avoid id-clashes.

Reply:
This sounds interesting. Really looking forward to discuss this more!

Reply:
This will be solved in #111